quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/KeycloakMain.java

```diff
@@ -140,45 +153,38 @@ private static boolean isFastStart(List<String> cliArgs) {
         return cliArgs.size() == 2 && cliArgs.get(0).equals(Start.NAME) && cliArgs.stream().anyMatch(OPTIMIZED_BUILD_OPTION_LONG::equals);
     }
 
-    public static void start(ExecutionExceptionHandler errorHandler, PrintWriter errStream) {
+    public static void start(Picocli picocli, ExecutionExceptionHandler errorHandler) {
         try {
             Quarkus.run(KeycloakMain.class, (exitCode, cause) -> {
                 if (cause != null) {
-                    errorHandler.error(errStream,
+                    errorHandler.error(picocli.getErrWriter(),
                             String.format("Failed to start server in (%s) mode", getKeycloakModeFromProfile(org.keycloak.common.util.Environment.getProfile())),
                             cause.getCause());
                 }
-
-                if (Environment.isDistribution()) {
-                    // assume that it is running the distribution
-                    // as we are replacing the default exit handler, we need to force exit
-                    System.exit(exitCode);
-                }
+                picocli.exit(exitCode);
             });
         } catch (Throwable cause) {
-            errorHandler.error(errStream,
+            errorHandler.error(picocli.getErrWriter(),
                     String.format("Unexpected error when starting the server in (%s) mode", getKeycloakModeFromProfile(org.keycloak.common.util.Environment.getProfile())),
                     cause.getCause());
-            System.exit(1);
         }
+        picocli.exit(CommandLine.ExitCode.SOFTWARE);
     }
 
     /**
      * Should be called after the server is fully initialized
      */
     @Override
     public int run(String... args) throws Exception {
-        int exitCode = ApplicationLifecycleManager.getExitCode();
-
         if (isTestLaunchMode() || isNonServerMode()) {
             // in test mode we exit immediately
             // we should be managing this behavior more dynamically depending on the tests requirements (short/long lived)
-            Quarkus.asyncExit(exitCode);
+            Quarkus.asyncExit(ApplicationLifecycleManager.getExitCode());
         } else {
             Quarkus.waitForExit();
         }
 
-        return exitCode;
+        return ApplicationLifecycleManager.getExitCode();
```

Comment:
Jus out of curiosity, why is this change needed? Why can't we store the exit code in a var?

Reply:
That's a potential bug, we don't want to store it because it could change while the waiting for exit.